ts-fold.el

```diff
@@ -392,6 +392,28 @@ Argument OFFSET can be used to tweak the final beginning and end position."
         (end (1- (tsc-node-end-position node))))
     (ts-fold--cons-add (cons beg end) offset)))
 
+(defun ts-fold-range-markers (node offset start-seq &optional end-seq)
+  "Returns the fold range for NODE with an OFFSET where the range
+starts at the end of the first occurence of the regular expression
+START-SEQ and ends at the end of the node or the last occurence of
+the optional regular expression LAST-SEQ.
+
+
+If no occurence is found for START-SEQ or END-SEQ or the
+occurences overlap, then the range returned is nil."
+  (when start-seq
+    (let ((beg (tsc-node-start-position node))
+          (end (tsc-node-end-position node)))
+      (save-excursion
+        (when-let* ((start-position (progn (goto-char beg)
+                                           (search-forward-regexp start-seq end t)))
+                    (end-position (if end-seq
+                                      (progn (goto-char end)
+                                             (search-backward-regexp end-seq beg t))
+                                    (1- end))))
+          (unless (> start-position end-position)
+            (ts-fold--cons-add (cons start-position end-position) offset)))))))
+
```

Comment:
Oops, sorry I should have mentioned we try not to add regex base solution to this package since we already using the tree-sitter so it's often better if you walk through tree-sitter AST! I should have documented in the README, I feel so sorry about this! 😓 

Check out the [tsc.el](https://github.com/emacs-tree-sitter/elisp-tree-sitter/blob/master/core/tsc.el), it has many useful functions to find node's parent, children, siblings, etc. You can also see other functions in `ts-fold.el` for examples!

Reply:
I originally had it set to just use `search-forward/backward` which works fine and I'm happy to switch it back to that. 

The rational I had for changing to regexp was:
1. Giving the user some flexibility since the AST doesn't always define nodes in the spots that may make sense to someone looking in.
2. You said elsewhere that you weren't looking to move into using more complex queries for defining folding nodes.

Neither of these are particularly strong sticking points with me. After all, if someone else needs it later, it can just become a new parser function 😛.

Reply:
Ah, no. I mean we should completely avoid search base solution, weather it's regex or not. Sorry I wasn't clear enough! 😅 Walk through AST is more robust and dedicated than the regular string/regex search.

> 1. Giving the user some flexibility since the AST doesn't always define nodes in the spots that may make sense to someone looking in.

The first argument will always be a valid node, so you should able to access what you need from that spot.

> 2. You said elsewhere that you weren't looking to move into using more complex queries for defining folding nodes

Yeah, I have tried implemented in #5 but the performance wasn't very good so I move on and not looking back once. 😅 

The goal of this package is to rely on tree-sitter as much as possible, or else it wouldn't use tree-sitter in the first place. I used to work on another regex base folding package [origami](https://github.com/gregsexton/origami.el), but it eventually collapsed and left many bugs due to their AST implementation.

Reply:
> I mean we should completely avoid search base solution

Given what you were asking for before, being able to support any amount of white space\new lines between the *interface* keyword and `{`, it simply isn't possible without a search of some sort. the tree-sitter AST nodes looks something like this (using `[` to represent boundaries of nodes):

```go
type Dog [interface {
    [Bark() (string, error)] // end of method_spec
    [Beg() (bool, error)] // end of method_spec
}] //end of interface_type
```

In theory the AST could be changed, but this is how it sits right now. 

I would argue that relying on the AST as much as possible is important and a very good thing, but having flexibility in place to fit cases where the AST doesn't quite solve it is also important. I think what you pointed out earlier that a definition should be in place that fits the desired fold regardless of format is a good thing.

Reply:
> Given what you were asking for before, being able to support any amount of white space\new lines between the interface keyword and {, it simply isn't possible without a search of some sort. the tree-sitter AST nodes looks something like this (using [ to represent boundaries of nodes):

It's possible but hidden by `tree-sitter-debug-mode`, here is the working implementation.

```elisp
;;; might place it ts-fold-util.el
(defun ts-fold--compare-type (node type)
  "Compare NODE's type to TYPE."
  (string= (ts-docstr-2-str (tsc-node-type node)) type))

(defun ts-fold-children (node)
  "Return children from NODE."
  (let (children)
    (dotimes (index (tsc-count-children node))
      (push (tsc-get-nth-child node index) children))
    (reverse children)))

(defun ts-fold-find-children (node type)
  "Search node TYPE from children; this return a list."
  (cl-remove-if-not (lambda (next) (ts-fold--compare-type next type))
                    (ts-fold-children node)))

;;; replace it in ts-fold.el
(defun ts-fold-range-markers (node offset start-seq &optional end-seq)
  "..."
  (let* ((beg-node (car (ts-fold-find-children node start-seq)))
         (end-node (car (ts-fold-find-children node end-seq)))
         (beg (1+ (tsc-node-start-position beg-node)))
         (end (1- (tsc-node-end-position end-node))))
    (ts-fold--cons-add (cons beg end) offset)))
```

*NOTE: Many functions copied from [ts-docstr](https://github.com/emacs-vs/ts-docstr)*

> I would argue that relying on the AST as much as possible is important and a very good thing, but having flexibility in place to fit cases where the AST doesn't quite solve it is also important. I think what you pointed out earlier that a definition should be in place that fits the desired fold regardless of format is a good thing.

Good point! I am trying balance between the two, if something doesn't fit then yes I will compromise when that comes! Thank you for you advice! 😄 

> Also, not needing to use origami is why I've been putting effort into this package recently 😆 It works fine most of the time, but it does tend to have bugs bubble up that can get frustrating to work around.

Oh nice! 😆 Didn't see that coming! 😁

Reply:
Sorry, got really busy today. Will try to look at this more in the next few days!

Reply:
I've gone through and think I understand how this is working now. Those anonymous nodes are quite annoying since they're hard to debug or see 😅. Since these sit on TS's CST instead of the AST, do you have any recommendations for how to see them in a given buffer? I have some ideas that I want to try for the implementation details, but there isn't much information about them in the [elisp doc site](https://emacs-tree-sitter.github.io/api/inspecting/). We should be able to get it working with this, but my biggest concern is that understanding or debugging these nodes could be a lot more work than the named nodes and so having a documented process in the README is important to add.

Note: I haven't looked at [ts-docstr](https://github.com/emacs-vs/ts-docstr) much yet since I was trying to understand this from first principles in TS's documentation. I'll look through that this weekend as I work through the 2-3 ideas that I have at the moment for this.

Reply:
> do you have any recommendations for how to see them in a given buffer?

No, sorry, I dig into the source code and see what `elisp-tree-sitter` provides then I play my way out. 😓 

> I have some ideas that I want to try for the implementation details, but there isn't much information about them in the [elisp doc site](https://emacs-tree-sitter.github.io/api/inspecting/). We should be able to get it working with this, but my biggest concern is that understanding or debugging these nodes could be a lot more work than the named nodes and so having a documented process in the README is important to add.

Yeah, I agree with you. It could make the emacs-tree-sitter development a lot easier. I don't know the project status from [elisp-tree-sitter](https://github.com/emacs-tree-sitter/elisp-tree-sitter), but it would be great if document it somewhere on the site. 👍 

> Note: I haven't looked at [ts-docstr](https://github.com/emacs-vs/ts-docstr) much yet since I was trying to understand this from first principles in TS's documentation. I'll look through that this weekend as I work through the 2-3 ideas that I have at the moment for this.

It's not necessary but that's my last project regarding elisp-tree-sitter. Hope the code make sense! 😅